File: packages/splunk_otel_flutter/splunk_otel_flutter/android/src/main/kotlin/com/splunk/rum/flutter/SplunkOtelFlutterPlugin.kt
/*
 * Copyright 2025 Splunk Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.splunk.rum.flutter


import android.app.Activity
import com.splunk.rum.flutter.extensions.toEndpointConfiguration
import com.splunk.rum.flutter.extensions.toGeneratedEndpointConfiguration
import com.splunk.rum.flutter.extensions.toGeneratedMutableAttributes
import com.splunk.rum.flutter.extensions.toGeneratedRecordingMaskList
import com.splunk.rum.flutter.extensions.toGeneratedRenderingMode
import com.splunk.rum.flutter.extensions.toGeneratedSessionReplayStatus
import com.splunk.rum.flutter.extensions.toGeneratedStatus
import com.splunk.rum.flutter.extensions.toGeneratedUserTrackingMode
import com.splunk.rum.flutter.extensions.toMutableAttributes
import com.splunk.rum.flutter.extensions.toRecordingMaskList
import com.splunk.rum.flutter.extensions.toRenderingMode
import com.splunk.rum.flutter.extensions.toUserTrackingMode
import com.splunk.rum.flutter.extensions.wrapIntoGeneratedMutableAttribute
import com.splunk.rum.integration.agent.api.AgentConfiguration
import com.splunk.rum.integration.agent.api.EndpointConfiguration
import com.splunk.rum.integration.agent.api.SplunkRum
import com.splunk.rum.integration.agent.api.session.SessionConfiguration
import com.splunk.rum.integration.agent.api.user.UserConfiguration
import com.splunk.rum.integration.agent.api.user.UserTrackingMode
import com.splunk.rum.integration.agent.common.attributes.MutableAttributes
import com.splunk.rum.integration.navigation.NavigationModuleConfiguration
import com.splunk.rum.integration.sessionreplay.extension.sessionReplay
import com.splunk.rum.integration.slowrendering.SlowRenderingModuleConfiguration
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import java.time.Duration

/** SplunkOtelFlutterPlugin */
class SplunkOtelFlutterPlugin :
    FlutterPlugin,
    ActivityAware,
    SplunkOtelFlutterHostApi {

    private var activity: Activity? = null

    // FlutterPlugin
    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        SplunkOtelFlutterHostApi.setUp(binding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    }

    // ActivityAware
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun install(
        agentConfiguration: GeneratedAgentConfiguration,
        navigationModuleConfiguration: GeneratedNavigationModuleConfiguration,
        slowRenderingModuleConfiguration: GeneratedSlowRenderingModuleConfiguration,
        callback: (Result<Unit>) -> Unit
    ) {

        val globalAttributes = agentConfiguration.globalAttributes?.toMutableAttributes()

        val agentConfiguration = AgentConfiguration(
            endpoint = agentConfiguration.endpoint.toEndpointConfiguration(),
            appName = agentConfiguration.appName,
            enableDebugLogging = agentConfiguration.enableDebugLogging ?: false,
            deploymentEnvironment = agentConfiguration.deploymentEnvironment,
            user = UserConfiguration(
                trackingMode = agentConfiguration.user?.trackingMode?.toUserTrackingMode() ?: UserTrackingMode.NO_TRACKING
            ),
            session = SessionConfiguration(agentConfiguration.session?.samplingRate ?: 1.0),
            globalAttributes = globalAttributes ?: MutableAttributes(),
            instrumentedProcessName = agentConfiguration.instrumentedProcessName,
            deferredUntilForeground = agentConfiguration.deferredUntilForeground ?: false,
        )

        val moduleConfigurations = arrayOf(
            NavigationModuleConfiguration(
                isEnabled = navigationModuleConfiguration.isEnabled,
                isAutomatedTrackingEnabled = navigationModuleConfiguration.isAutomatedTrackingEnabled,
            ),
            SlowRenderingModuleConfiguration(
                isEnabled = slowRenderingModuleConfiguration.isEnabled,
                interval = Duration.ofMillis((slowRenderingModuleConfiguration.intervalMillis))
            )
        )

        try {
            if(activity == null){
                callback(Result.failure(FlutterError("INSTALL_FAILED", "Activity not available yet")))

                return
            }

            SplunkRum.install(
                application = activity!!.application,
                agentConfiguration = agentConfiguration,
                moduleConfigurations = moduleConfigurations
            )
        } catch (e: Exception) {
            callback(Result.failure(FlutterError("INSTALL_FAILED", e.message)))

            return
        }

        callback(Result.success(Unit))
    }

    // Session replay

    override fun sessionReplayStart(callback: (Result<Unit>) -> Unit) {
        SplunkRum.instance.sessionReplay.start();

        callback(Result.success(Unit))
    }

    override fun sessionReplayStop(callback: (Result<Unit>) -> Unit) {
        SplunkRum.instance.sessionReplay.stop();

        callback(Result.success(Unit))
    }

    override fun sessionReplayStateGetStatus(callback: (Result<GeneratedSessionReplayStatus>) -> Unit) {
        val status = SplunkRum.instance.sessionReplay.state.status

        callback(Result.success(status.toGeneratedSessionReplayStatus()))
    }

    override fun sessionReplayStateGetRenderingMode(callback: (Result<GeneratedRenderingMode>) -> Unit) {
        val renderingMode = SplunkRum.instance.sessionReplay.state.renderingMode

        callback(Result.success(renderingMode.toGeneratedRenderingMode()))
    }

    override fun sessionReplayPreferencesGetRenderingMode(callback: (Result<GeneratedRenderingMode?>) -> Unit) {
        val renderingMode = SplunkRum.instance.sessionReplay.preferences.renderingMode

        callback(Result.success(renderingMode?.toGeneratedRenderingMode()))
    }

    override fun sessionReplayPreferencesSetRenderingMode(
        renderingMode: GeneratedRenderingMode?,
        callback: (Result<Unit>) -> Unit
    ) {
        SplunkRum.instance.sessionReplay.preferences.renderingMode = renderingMode?.toRenderingMode()

        callback(Result.success(Unit))
    }

    override fun sessionReplayGetRecordingMask(callback: (Result<GeneratedRecordingMaskList?>) -> Unit) {
        val recordingMask = SplunkRum.instance.sessionReplay.recordingMask

        callback(Result.success(recordingMask?.toGeneratedRecordingMaskList()))
    }

    override fun sessionReplaySetRecordingMask(
        recordingMask: GeneratedRecordingMaskList?,
        callback: (Result<Unit>) -> Unit
    ) {
        SplunkRum.instance.sessionReplay.recordingMask = recordingMask?.toRecordingMaskList()

        callback(Result.success(Unit))
    }

    // State

    override fun stateGetAppName(callback: (Result<String>) -> Unit) {
        val appName = SplunkRum.instance.state.appName

        callback(Result.success(appName))
    }

    override fun stateGetAppVersion(callback: (Result<String>) -> Unit) {
        val appVersion = SplunkRum.instance.state.appVersion

        callback(Result.success(appVersion))
    }

    override fun stateGetStatus(callback: (Result<GeneratedStatus>) -> Unit) {
        val status = SplunkRum.instance.state.status

        callback(Result.success(status.toGeneratedStatus()))
    }

    override fun stateGetEndpointConfiguration(callback: (Result<GeneratedEndpointConfiguration>) -> Unit) {
        val endpointConfiguration = SplunkRum.instance.state.endpointConfiguration

        callback(Result.success(endpointConfiguration.toGeneratedEndpointConfiguration()))
    }

    override fun stateGetDeploymentEnvironment(callback: (Result<String>) -> Unit) {
        val deploymentEnvironment = SplunkRum.instance.state.deploymentEnvironment

        callback(Result.success(deploymentEnvironment))
    }

    override fun stateGetIsDebugLoggingEnabled(callback: (Result<Boolean>) -> Unit) {
        val isDebugLoggingEnabled = SplunkRum.instance.state.isDebugLoggingEnabled

        callback(Result.success(isDebugLoggingEnabled))
    }

    override fun stateGetInstrumentedProcessName(callback: (Result<String?>) -> Unit) {
        val instrumentedProcessName = SplunkRum.instance.state.instrumentedProcessName

        callback(Result.success(instrumentedProcessName))
    }

    override fun stateGetDeferredUntilForeground(callback: (Result<Boolean>) -> Unit) {
        val deferredUntilForeground = SplunkRum.instance.state.deferredUntilForeground

        callback(Result.success(deferredUntilForeground))
    }

    // Session

    override fun sessionStateGetId(callback: (Result<String>) -> Unit) {
        val sessionId = SplunkRum.instance.session.state.id

        callback(Result.success(sessionId))
    }

    override fun sessionStateGetSamplingRate(callback: (Result<Double>) -> Unit) {
        val samplingRate = SplunkRum.instance.session.state.samplingRate

        callback(Result.success(samplingRate))
    }

    // User

    override fun userStateGetUserTrackingMode(callback: (Result<GeneratedUserTrackingMode>) -> Unit) {
        val trackingMode = SplunkRum.instance.user.state.trackingMode

        callback(Result.success(trackingMode.toGeneratedUserTrackingMode()))
    }

    override fun userPreferencesGetUserTrackingMode(callback: (Result<GeneratedUserTrackingMode?>) -> Unit) {
        val trackingMode = SplunkRum.instance.user.preferences.trackingMode

        callback(Result.success(trackingMode?.toGeneratedUserTrackingMode()))
    }

    override fun userPreferencesSetUserTrackingMode(
        trackingMode: GeneratedUserTrackingMode,
        callback: (Result<Unit>) -> Unit
    ) {
        SplunkRum.instance.user.preferences.trackingMode = trackingMode.toUserTrackingMode()

        callback(Result.success(Unit))
    }

    // Global attributes

    override fun globalAttributesGet(key: String, callback: (Result<Any?>) -> Unit) {
        val attribute = SplunkRum.instance.globalAttributes.get<Any?>(key)
        val convertedAttribute = attribute?.wrapIntoGeneratedMutableAttribute()

        callback(Result.success(convertedAttribute))
    }

    override fun globalAttributesGetAll(callback: (Result<GeneratedMutableAttributes?>) -> Unit) {
        val attributes = SplunkRum.instance.globalAttributes.getAll()
        val convertedAttributes = attributes.toGeneratedMutableAttributes()

        callback(Result.success(convertedAttributes))
    }

    override fun globalAttributesRemove(key: String, callback: (Result<Unit>) -> Unit) {
        SplunkRum.instance.globalAttributes.remove(key)

        callback(Result.success(Unit))
    }

    override fun globalAttributesRemoveAll(callback: (Result<Unit>) -> Unit) {
        SplunkRum.instance.globalAttributes.removeAll()

        callback(Result.success(Unit))
    }

    override fun globalAttributesContains(key: String, callback: (Result<Boolean>) -> Unit) {
        val doesContain = SplunkRum.instance.globalAttributes.contains(key)

        callback(Result.success(doesContain))
    }

    override fun globalAttributesSetString(
        key: String,
        value: String,
        callback: (Result<Unit>) -> Unit
    ) {
        SplunkRum.instance.globalAttributes[key] = value

        callback(Result.success(Unit))
    }

    override fun globalAttributesSetInt(key: String, value: Long, callback: (Result<Unit>) -> Unit) {
        SplunkRum.instance.globalAttributes[key] = value

        callback(Result.success(Unit))
    }

    override fun globalAttributesSetDouble(
        key: String,
        value: Double,
        callback: (Result<Unit>) -> Unit
    ) {
        SplunkRum.instance.globalAttributes[key] = value

        callback(Result.success(Unit))
    }

    override fun globalAttributesSetBool(
        key: String,
        value: Boolean,
        callback: (Result<Unit>) -> Unit
    ) {
        SplunkRum.instance.globalAttributes[key] = value

        callback(Result.success(Unit))
    }

    override fun globalAttributesSetStringList(
        key: String,
        value: List<String>,
        callback: (Result<Unit>) -> Unit
    ) {
        SplunkRum.instance.globalAttributes.update {
            put(key, *value.toTypedArray())
        }

        callback(Result.success(Unit))
    }

    override fun globalAttributesSetIntList(
        key: String,
        value: List<Long>,
        callback: (Result<Unit>) -> Unit
    ) {
        SplunkRum.instance.globalAttributes.update {
            put(key, *value.toLongArray())
        }

        callback(Result.success(Unit))
    }

    override fun globalAttributesSetDoubleList(
        key: String,
        value: List<Double>,
        callback: (Result<Unit>) -> Unit
    ) {
        SplunkRum.instance.globalAttributes.update {
            put(key, *value.toDoubleArray())
        }

        callback(Result.success(Unit))
    }

    override fun globalAttributesSetBoolList(
        key: String,
        value: List<Boolean>,
        callback: (Result<Unit>) -> Unit
    ) {
        SplunkRum.instance.globalAttributes.update {
            put(key, *value.toBooleanArray())
        }

        callback(Result.success(Unit))
    }

    override fun globalAttributesSetAll(
        value: GeneratedMutableAttributes,
        callback: (Result<Unit>) -> Unit
    ) {
        try {
            SplunkRum.instance.globalAttributes.setAll(value.toMutableAttributes())
        }catch (e:Exception)
        {
            callback(Result.failure(FlutterError("SET_ALL_FAILED", e.message)))

            return
        }

        callback(Result.success(Unit))
    }
}

File: packages/splunk_otel_flutter/splunk_otel_flutter/ios/splunk_otel_flutter/Sources/splunk_otel_flutter/SplunkOtelFlutterPlugin.swift
//
/*
 Copyright 2025 Splunk Inc.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

import Flutter
import UIKit
@_spi(SplunkInternal) import SplunkAgent
import OpenTelemetryApi
import SplunkSlowFrameDetector
import SplunkNavigation
import SplunkAppStart
import SplunkNetwork
import SplunkInteractions
import SplunkNetworkMonitor
import SplunkCrashReports

extension FlutterError: Error {}

public class SplunkOtelFlutterPlugin: NSObject, FlutterPlugin, SplunkOtelFlutterHostApi {
    
    private var didFinishLaunchingAt: Date?
    private var willEnterForegroundAt: Date?
    private var workflowSpans: [Int64: (span: Span, startTime: Int64)] = [:]
    private var workflowHandleCounter: Int64 = 0
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = SplunkOtelFlutterPlugin()
        
        // Observe lifecycle notifications
           NotificationCenter.default.addObserver(
               instance,
               selector: #selector(onDidBecomeActive),
               name: UIApplication.didBecomeActiveNotification,
               object: nil
           )

           NotificationCenter.default.addObserver(
               instance,
               selector: #selector(onDidFinishLaunchingNotification),
               name: UIApplication.didFinishLaunchingNotification,
               object: nil
           )

           NotificationCenter.default.addObserver(
               instance,
               selector: #selector(onWillEnterForeground),
               name: UIApplication.willEnterForegroundNotification,
               object: nil
           )
        
        SplunkOtelFlutterHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    // MARK: - Notification handlers
    @objc private func onDidFinishLaunchingNotification(_ note: Notification) {
        let launchOptions = note.userInfo as? [UIApplication.LaunchOptionsKey: Any]
        handleDidFinishLaunching(launchOptions: launchOptions)
    }
    
    @objc private func onDidBecomeActive(_ note: Notification) {
        handleDidBecomeActive()
    }
    
    @objc private func onWillEnterForeground(_ note: Notification) {
          handleWillEnterForeground()
      }
    
    // MARK: - Internals
    private func handleDidFinishLaunching(launchOptions: [UIApplication.LaunchOptionsKey: Any]?) {
        didFinishLaunchingAt = Date()
    }

    private func handleWillEnterForeground() {
        // Mark the moment the app transitions toward foreground (warm start path)
        willEnterForegroundAt = Date()
    }
    
 

    private func handleDidBecomeActive() {
        // The "active" moment (common for both cold and warm starts)
        let becameActiveAt = Date()

        // Send app start timing to Splunk RUM.
        // - didFinishLaunchingAt: non-nil indicates a cold start path
        // - willEnterForegroundAt: non-nil indicates a warm start path
        // Splunk RUM will infer the correct variant based on which timestamps are provided.
        SplunkRum.shared.appStart.track(
            didBecomeActive: becameActiveAt,
            didFinishLaunching: didFinishLaunchingAt,
            willEnterForeground: willEnterForegroundAt
        )


        // Clear foreground marker so a later activation (e.g., another background->foreground) is measured correctly.
        willEnterForegroundAt = nil
    }
    
    func install(agentConfiguration: GeneratedAgentConfiguration,
                 navigationModuleConfiguration: GeneratedNavigationModuleConfiguration?,
                 slowRenderingModuleConfiguration: GeneratedSlowRenderingModuleConfiguration?,
                 crashReportsModuleConfiguration: GeneratedCrashReportsModuleConfiguration?,
                 interactionsModuleConfiguration: GeneratedInteractionsModuleConfiguration?,
                 networkMonitorModuleConfiguration: GeneratedNetworkMonitorModuleConfiguration?,
                 applicationLifecycleModuleConfiguration: GeneratedApplicationLifecycleModuleConfiguration?,
                 // Android-only
                 anrModuleConfiguration: GeneratedAnrModuleConfiguration?, 
                 httpUrlModuleConfiguration: GeneratedHttpUrlModuleConfiguration?, 
                 okHttp3AutoModuleConfiguration: GeneratedOkHttp3AutoModuleConfiguration?, 
                 // iOS-only
                 networkInstrumentationModuleConfiguration: GeneratedNetworkInstrumentationModuleConfiguration?,
                 completion: @escaping (Result<Void, any Error>) -> Void) {
        
        
        // should be always not nil
        guard let endpointConfiguration = agentConfiguration.endpoint.toEndpointConfiguration() else {
            completion(
                .failure(
                    FlutterError(
                        code: "INSTALL_FAILED",
                        message: "Endpoint configuration - Invalid parameter combination",
                        details: nil
                    )
                )
            )
            return
        }
        
     
        let mergedGlobalAttributes =
            agentConfiguration.globalAttributes?.toMutableAttributes() ?? MutableAttributes()
        //TODO move to Resources later
        mergedGlobalAttributes.setString(
            SplunkRumFlutterTelemetryVersions.rumSdkFlutterVersion,
            for: SplunkRumFlutterTelemetryAttributeKeys.rumSdkFlutterVersion)

        let agentConfig = AgentConfiguration(
            endpoint: endpointConfiguration,
            appName: agentConfiguration.appName,
            deploymentEnvironment: agentConfiguration.deploymentEnvironment,
        )
            .appVersion(agentConfiguration.appVersion ?? "")
            .enableDebugLogging(agentConfiguration.enableDebugLogging ?? false)
            .globalAttributes(mergedGlobalAttributes)
            .sessionConfiguration(SessionConfiguration(samplingRate: agentConfiguration.session?.samplingRate ?? 1.0))
            .userConfiguration(UserConfiguration(trackingMode: agentConfiguration.user?.trackingMode == .anonymousTracking ? .anonymousTracking : .noTracking))
        do {
            let moduleConfigurations: [Any] = [
                   slowRenderingModuleConfiguration.map {
                       SlowFrameDetectorConfiguration(isEnabled: $0.isEnabled)
                   },
                   navigationModuleConfiguration.map {
                       NavigationConfiguration(
                           isEnabled: $0.isEnabled,
                           enableAutomatedTracking: $0.isAutomatedTrackingEnabled
                       )
                   },
                   networkInstrumentationModuleConfiguration.map {
                       let ignoredUrls = (try? $0.ignoreURLs.toIgnoreURLs()) ?? IgnoreURLs()
                       return NetworkInstrumentationConfiguration(
                           isEnabled: $0.isEnabled,
                           ignoreURLs: ignoredUrls
                       )
                   },
                   crashReportsModuleConfiguration.map {
                       CrashReportsConfiguration(
                           isEnabled: $0.isEnabled,
                       )
                   },
                   interactionsModuleConfiguration.map {
                       InteractionsConfiguration(
                           isEnabled: $0.isEnabled,
                       )
                   },
                   networkMonitorModuleConfiguration.map {
                       NetworkMonitorConfiguration(
                           isEnabled: $0.isEnabled,
                       )
                   },
               ].compactMap { $0 }// removes nils automatically
            
            let agent = try SplunkRum.install(with: agentConfig, moduleConfigurations: moduleConfigurations)
          
            handleDidBecomeActive()
            
            completion(.success(()))
        } catch {
            completion(.failure(
                FlutterError(
                    code: "INSTALL_FAILED",
                    message: error.localizedDescription,
                    details: nil
                )
            ))
        }
    }
    
    // MARK: - State
    
    func stateGetAppName(completion: @escaping (Result<String, any Error>) -> Void) {
        let appName = SplunkRum.shared.state.appName
        
        completion(.success(appName))
    }
    
    func stateGetAppVersion(completion: @escaping (Result<String, any Error>) -> Void) {
        let appVersion = SplunkRum.shared.state.appVersion
        
        completion(.success(appVersion))
    }
    
    func stateGetStatus(completion: @escaping (Result<GeneratedStatus, any Error>) -> Void) {
        let status = SplunkRum.shared.state.status
        
        completion(.success(status.toGeneratedStatus()))
    }
    
    func stateGetEndpointConfiguration(completion: @escaping (Result<GeneratedEndpointConfiguration, any Error>) -> Void) {
        let endpointConfiguration = SplunkRum.shared.state.endpointConfiguration
        
        completion(.success(endpointConfiguration.toGeneratedEndpointConfiguration()))
    }
    
    func stateGetDeploymentEnvironment(completion: @escaping (Result<String, any Error>) -> Void) {
        let environment = SplunkRum.shared.state.deploymentEnvironment
        
        completion(.success(environment))
    }
    
    func stateGetIsDebugLoggingEnabled(completion: @escaping (Result<Bool, any Error>) -> Void) {
        let isDebugLoggingEnabled = SplunkRum.shared.state.isDebugLoggingEnabled
        
        completion(.success(isDebugLoggingEnabled))
    }
    
    func stateGetInstrumentedProcessName(completion: @escaping (Result<String?, any Error>) -> Void) {
        completion(.success(nil))
    }
    
    func stateGetDeferredUntilForeground(completion: @escaping (Result<Bool, any Error>) -> Void) {
        completion(.success(false))
    }
    
    // MARK: - Preferences
    
    func preferencesGetEndpointConfiguration(completion: @escaping (Result<GeneratedEndpointConfiguration?, any Error>) -> Void) {
        //TODO
        
        //let endpointConfiguration = SplunkRum.shared.
        
        //completion(.success(endpointConfiguration?.toGeneratedEndpointConfiguration()))
    }
    
    // MARK: - Session
    
    func sessionStateGetId(completion: @escaping (Result<String, any Error>) -> Void) {
        let sessionId = SplunkRum.shared.session.state.id
        
        completion(.success(sessionId))
    }
    
    func sessionStateGetSamplingRate(completion: @escaping (Result<Double, any Error>) -> Void) {
        let samplingRate = SplunkRum.shared.session.state.samplingRate
        
        completion(.success(samplingRate))
    }
    
    // MARK: - User
    
    func userStateGetUserTrackingMode(completion: @escaping (Result<GeneratedUserTrackingMode, any Error>) -> Void) {
        let trackingMode = SplunkRum.shared.user.state.trackingMode
        
        completion(.success(trackingMode.toGeneratedUserTrackingMode()))
    }
    
    func userPreferencesGetUserTrackingMode(completion: @escaping (Result<GeneratedUserTrackingMode?, any Error>) -> Void) {
        let trackingMode = SplunkRum.shared.user.preferences.trackingMode
        
        completion(.success(trackingMode.toGeneratedUserTrackingMode()))
    }
    
    func userPreferencesSetUserTrackingMode(trackingMode: GeneratedUserTrackingMode, completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.user.preferences.trackingMode = trackingMode.toUserTrackingMode()
        
        completion(.success(()))
    }
    
    // MARK: - Global Attributes
    
    func globalAttributesGet(key: String, completion: @escaping (Result<Any?, any Error>) -> Void) {
        let attribute = SplunkRum.shared.globalAttributes.getValue(for: key)
        let convertedAttribute = attribute?.toGeneratedWrapper()
        
        completion(.success(convertedAttribute))
    }
    
    func globalAttributesGetAll(completion: @escaping (Result<GeneratedMutableAttributes?, Error>) -> Void) {
        let attributes = SplunkRum.shared.globalAttributes.getAll()
        
        var converted: [String: Any?] = [:]
        for (key, value) in attributes {
            converted[key] = value.toGeneratedWrapper()
        }
        let generated = GeneratedMutableAttributes(attributes: converted)
        
        completion(.success(generated))
    }
    
    func globalAttributesRemove(key: String, completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.globalAttributes.remove(for: key)
        
        completion(.success(()))
    }
    
    func globalAttributesRemoveAll(completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.globalAttributes.removeAll()
        
        completion(.success(()))
    }
    
    func globalAttributesContains(key: String, completion: @escaping (Result<Bool, any Error>) -> Void) {
        let doesContain = SplunkRum.shared.globalAttributes.contains(key: key)
        
        completion(.success(doesContain))
    }
    
    func globalAttributesSetString(key: String, value: String, completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.globalAttributes.setString(value, for: key)
        
        completion(.success(()))
    }
    
    func globalAttributesSetInt(key: String, value: Int64, completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.globalAttributes.setInt(Int(value), for: key)
        
        completion(.success(()))
    }
    
    func globalAttributesSetDouble(key: String, value: Double, completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.globalAttributes.setDouble(value, for: key)
        
        completion(.success(()))
    }
    
    func globalAttributesSetBool(key: String, value: Bool, completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.globalAttributes.setBool(value, for: key)
        
        completion(.success(()))
    }
    
    func globalAttributesSetStringList(key: String, value: [String], completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.globalAttributes.setArray(AttributeArray.fromStringArray(value), for: key)
        
        completion(.success(()))
    }
    
    func globalAttributesSetIntList(key: String, value: [Int64], completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.globalAttributes.setArray(AttributeArray.fromIntArray(value), for: key)
        
        completion(.success(()))
    }
    
    func globalAttributesSetDoubleList(key: String, value: [Double], completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.globalAttributes.setArray(AttributeArray.fromDoubleArray(value), for: key)
        
        completion(.success(()))
    }
    
    func globalAttributesSetBoolList(key: String, value: [Bool], completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.globalAttributes.setArray(AttributeArray.fromBoolArray(value), for: key)
        
        completion(.success(()))
    }
    
    func globalAttributesSetAll(value: GeneratedMutableAttributes, completion: @escaping (Result<Void, any Error>) -> Void) {
        let attributes = SplunkRum.shared.globalAttributes
        
        value.attributes.forEach { (key, wrapped) in
            switch wrapped {
            case let v as GeneratedMutableAttributeInt:
                attributes.setInt(Int(v.value),for: key)
                
            case let v as GeneratedMutableAttributeDouble:
                attributes.setDouble(v.value,for: key)
                
            case let v as GeneratedMutableAttributeString:
                attributes.setString(v.value,for: key)
                
            case let v as GeneratedMutableAttributeBool:
                attributes.setBool(v.value,for: key)
                
            case let v as GeneratedMutableAttributeListInt:
                attributes.setArray(AttributeArray.fromIntArray(v.value),for: key)
                
            case let v as GeneratedMutableAttributeListDouble:
                attributes.setArray(AttributeArray.fromDoubleArray(v.value),for: key)
                
            case let v as GeneratedMutableAttributeListString:
                attributes.setArray(AttributeArray.fromStringArray(v.value),for: key)
                
            case let v as GeneratedMutableAttributeListBool:
                attributes.setArray(AttributeArray.fromBoolArray(v.value),for: key)
                
            case .none:
                break
                
            default:
                completion(.failure(FlutterError(
                    code: "SET_ALL_FAILED",
                    message:     "Unsupported GeneratedMutableAttribute type for key \(key): \(String(describing: wrapped))",
                    details: nil
                )
                ))
                return
            }
        }
        
        completion(.success(()))
    }
    
    // MARK: - Custom tracking
    
    func customTrackingTrackCustomEvent(name: String, attributes: GeneratedMutableAttributes, completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.customTracking.trackCustomEvent(name, attributes.toMutableAttributes())
        
        completion(.success(()))
    }
    
    func customTrackingStartWorkflow(workflowName: String, completion: @escaping (Result<Int64, any Error>) -> Void) {
        // Start the workflow and generate a unique handle
        let span = SplunkRum.shared.customTracking.trackWorkflow(workflowName)
        let startTime = Int64(Date().timeIntervalSince1970 * 1000)
        
        workflowHandleCounter += 1
        let handle = workflowHandleCounter
        
        workflowSpans[handle] = (span: span, startTime: startTime)
        
        completion(.success(handle))
    }
    
    func customTrackingEndWorkflow(handle: Int64, completion: @escaping (Result<Void, any Error>) -> Void) {
        guard let workflow = workflowSpans[handle] else {
            completion(.failure(NSError(domain: "SplunkOtelFlutter", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid workflow handle"])))
            return
        }
        
        // End the workflow
        let endTime = Int64(Date().timeIntervalSince1970 * 1000)
        
        workflow.span.setAttribute(key: "workflow.start.time", value: AttributeValue.int(Int(workflow.startTime)))
        workflow.span.setAttribute(key: "workflow.end.time", value: AttributeValue.int(Int(endTime)))
        workflow.span.end()
        
        workflowSpans.removeValue(forKey: handle)
        
        completion(.success(()))
    }
    
    // MARK: - Navigation
    
    func navigationTrack(screenName: String, completion: @escaping (Result<Void, any Error>) -> Void) {
        SplunkRum.shared.navigation.track(screen: screenName)
        
        completion(.success(()))
    }
}
